### PR TITLE
Introduce SolverId/SolverInstallPlan as solver output.

### PR DESCRIPTION
**Pull request by ezyang**
Just the last patch in the series is new. 

Currently, dependency solving immediately produces an 'InstallPlan'
which is then consumed by cabal install, or elaborated into
an 'ElaboratedInstallPlan' for cabal new-build.  However, this
translation is awkward, because the dependency solver knows nothing
about 'UnitId's, yet an 'InstallPlan' must indexed by 'UnitId's.
So there's a bit of faffing around to generate a "fake" unit id
to satisfy the interface, and then eventually correct it to the
right one.

So this patch starts moving us in a better direction, by introducing
a 'SolverInstallPlan', 'SolverPackage' and 'SolverId', intended
to be generated by the solver.  Then 'configureInstallPlan' or
'elaborateInstallPlan' elaborate this representation into the
representation needed by the destination.

The next step will be to generate the 'UnitId' during
'configureInstallPlan', and then we can get rid of the fake map
(so only Solver data types generate a fake identity, which
is only temporary until we generate 'UnitId's.)

### CODE REVIEW

File: cabal-install/Distribution/Client/Types.hs [hunk 1]
```diff
@@ -165,6 +165,58 @@ instance HasUnitId (ConfiguredPackage loc) where
...
+        solverPkgDeps :: ComponentDeps [SolverId]
+    }
+  deriving (Eq, Show, Generic)
+
+instance Binary loc => Binary (SolverPackage loc)
+
+instance Package (SolverPackage loc) where
+  packageId = packageId . solverPkgSource
+
+-- | This is a minor hack as 'PackageIndex' is assumes keys are
```
Comment (23Skidoo) [inline cabal-install/Distribution/Client/Types.hs]: s/is assumes/assumes/

**Comment by 23Skidoo** [comment]
Typo in commit message: "must indexed".

**Comment by 23Skidoo** [comment]
Looks OK. 

/cc @dcoutts

**Comment by ezyang** [comment]
Going to reopen this PR.

**Pull request closed without merging**